README.md

```diff
@@ -1,14 +1,124 @@
 Reverse Tunneling Dialer
 ========================
 
-Client makes an outbound connection to a server.  The server can now do net.Dial from the
-server that will actually do a net.Dial on the client and pipe all bytes back and forth.
+remotedialer creates a two-way connection between two entities which allows
+services running on localhost or behind a NAT or firewall to be accessible by a
+remote server.
 
-Fun times!
+Architecture
+------------
 
-Refer to [`server/`](server/) and [`client/`](client/) how to use.  Or don't.... This framework can hurt your head
-trying to conceptualize.
+### Abstractions
 
-See also:
+remotedialer consists of structs that organize and abstract a TCP connection
+between a server and a client. Both client and server use ``Session``s, which
+provides a means to make a network connection to a remote endpoint and keep
+track of incoming connections. A server uses the ``Server`` object which
+contains a ``sessionManager`` which governs one or more ``Session``s, while a
+client creates a ``Session`` directly. A ``connection`` implements the
+``io.Reader`` and ``io.WriteCloser`` interfaces, so it can be read from and
+written to directly. The ``connection``'s internal ``readBuffer`` monitors the
+size of the data it is carrying and uses ``backPressure`` to pause incoming
```

Comment:
In the server there is a BackPressure mechanism per connection but there is nothing preventing a large spike in memory due to multiple buffers being filled at once right?

Reply:
Not that I can see. There's one backPressure for readBuffer and one readBuffer per connection and there's no practical limit to connections (the connection map has int64 keys so technically the limit is 2^64).